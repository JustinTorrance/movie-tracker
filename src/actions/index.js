export const loadMovies = (movies) => ({
  type: 'LOAD_MOVIES',
  movies
})

export const signIn = (user) => ({
  type: 'SIGN_IN',
  user
})

export const signOut = () => ({
  type: 'SIGN_OUT'
})

export const loading = (bool) => ({
  type: 'IS_LOADING',
  bool
})

export const catchError = (bool) => ({
  type: 'IS_ERROR',
  bool
})
