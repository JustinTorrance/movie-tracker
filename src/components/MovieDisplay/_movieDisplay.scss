.movie-tracker {
  color: $color-blue;
  font-family: $title-font;
  font-size: 80px;
  margin: 0px 30px 25px;
}

.main-logo {
  width: 50px;
  height: 50px;
  position: relative;
  bottom: -4px;
}

.user-greeting {
  color: white;
  right: 44px;
  top: 30px;
  position: absolute;
  padding: 0;
  margin: 0;
  font-size: 1.2rem;
}

.sign-out-btn, .toggle-favorites-btn {
  background-color: #2D333B;
  border: 2px solid $color-blue;
  border-radius: 18px;
  color: #B7B7B7;
  position: absolute;
  font-family: $regular-font;
  font-size: 1.1em;
  padding: 6px 8px;
  top: 25px;
  transition: all .3s ease;

  &:hover {
    cursor: pointer;
    color: white;
    background-color: rgb(22, 131, 194);
    border-color: #2D333B
  }
}

.sign-out-btn {
  right: 165px;
}

.toggle-favorites-btn {
  right: 265px;
}

.movie-container-label {
  font-family: $regular-font;
  font-weight: 300;
  color: white;
  font-size: 3em;
  margin: 5px 30px;
}

.movie-container-underline { 
  background-color: white;
  height: 2px;
  width: 95%;
  margin: 10px 30px;
}

.movies-container {
  display: grid;
  grid-template-columns: repeat(auto-fit, 254px);
  grid-gap: 20px;
  justify-content: center;
  margin: 30px auto;
  width: calc(100% - 60px)
}