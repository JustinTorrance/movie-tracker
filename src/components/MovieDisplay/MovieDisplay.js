import React, { Component } from 'react'
import { connect } from 'react-redux'
import MovieCard from '../MovieCard/MovieCard'
import { uid } from 'react-uid'
import { signOut } from '../../actions/index'
import { PropTypes } from 'prop-types'
import { getFavorites } from '../../utils/apiCalls.js'
import Loading from '../Loading/Loading'
import { Link } from 'react-router-dom'

export class MovieDisplay extends Component {
  constructor() {
    super()
    this.state = {
      favorites: '',
      favoriteMovies: []
    }
  }

  componentDidMount = async () => {
    const favoriteMovies = await getFavorites(this.props.user_id)
    this.setState({favoriteMovies: favoriteMovies.data})
  }

  async toggleFavorites() {
    let current = ''
    const favoriteMovies = await getFavorites(this.props.user_id)
    if (!this.state.favorites) {
      current = 'favorites'
    }
    this.setState({
      favorites: current,
      favoriteMovies: favoriteMovies.data
    })
  }

  render() {
    const { movies, signOut, isLoading, user } = this.props
    let renderedMovies = movies
    if(this.state.favorites) {
      renderedMovies = this.state.favoriteMovies
      var title = 'Favorites'
      var buttonText = 'trending'
    } else {
      var title = 'Trending'
      var buttonText = 'favorites'
    }
    if(isLoading) {
      return (<Loading />)
    }
    return(
      <section className="movies-section-container">
        <header className='movies-header'>
          <h1 className="movie-tracker">  
            movie
            <img src='./images/movie_roll.svg' alt="movie roll" className='main-logo'></img>
            tracker
          </h1>
          <p className='user-greeting'>Hello, {user.name}</p>
          <Link to={'/login'}>
            <button className='sign-out-btn' onClick={signOut}>sign out</button>
          </Link>
          <Link to={`/${buttonText}`}>
            <button className='toggle-favorites-btn' onClick={() => this.toggleFavorites()}>{ buttonText}</button>
          </Link>
        </header>
        <h2 className='movie-container-label'>{title}</h2>
        <div className='movie-container-underline'></div>
        <div className="movies-container">{renderedMovies.map((movie) => {
          return (
            <MovieCard movie={movie} key={uid(movie)} reRender={this.componentDidMount}/>
          )
        })}
        {
          renderedMovies.length === 0 && <h2>You Have No Favorites</h2>
        }
        </div>
      </section>
    )
    }
  }

export const mapStateToProps = (state) => ({
  movies: state.movies,
  user_id: state.user.id,
  isLoading: state.isLoading,
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  signOut: () => dispatch(signOut())
})

MovieDisplay.propTypes = {
  movies: PropTypes.array.isRequired,
  signOut: PropTypes.func.isRequired,
  user_id: PropTypes.number.isRequired,
  isLoading: PropTypes.bool,
  user: PropTypes.object.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(MovieDisplay)