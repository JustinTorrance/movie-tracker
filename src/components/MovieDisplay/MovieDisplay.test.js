import React from 'react'
import { shallow } from 'enzyme' 
import { MovieDisplay } from './MovieDisplay'
import { signOut } from '../../actions/index'
import { mapStateToProps, mapDispatchToProps } from './MovieDisplay'

describe('MovieDisplay', () => {
  const mockMovie = {
    id: 1,
    title: 'Titanic',
    year: 1999,
    rating: 9,
    posterPic: 'https://image.tmdb.org/t/p/w500///',
    backdropPic: 'https://image.tmdb.org/t/p/w500///',
    overview: 'great movie',
    genres: 'drama',
    runtime: 50
  }
  const mockSignOut = jest.fn()

  it('Should match the snapshot', () => {
    const wrapper = shallow(<MovieDisplay movies={[mockMovie]} signOut={mockSignOut}/>)
    
    expect(wrapper).toMatchSnapshot()
  })

  describe('MapStateToProps', () => {
    it('Should create a object with a movies array', () => {
      const mockState = {
        movies: [mockMovie],
      }

      const expected = {movies: [mockMovie]}

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('MapDispatchToProps', () => {
    it('Calls dispatch with a signOut action', () => {
      const mockDispatch = jest.fn()
      const actionToDispatch = signOut()
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.signOut()
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})