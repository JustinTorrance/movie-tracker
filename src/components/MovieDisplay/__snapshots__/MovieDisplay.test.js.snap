// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDisplay Should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieDisplay
    isLoading={false}
    loadMovies={
      [MockFunction] {
        "calls": Array [
          Array [
            "https://api.themoviedb.org/3/movie/popular?api_key=da90047b6c1d3526d4b04666a1b64a0d&language=en-US&page=1&region=US",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    movies={
      Array [
        Object {
          "backdropPic": "https://image.tmdb.org/t/p/w500///",
          "genres": "drama",
          "id": 1,
          "overview": "great movie",
          "posterPic": "https://image.tmdb.org/t/p/w500///",
          "rating": 9,
          "runtime": 50,
          "title": "Titanic",
          "year": 1999,
        },
      ]
    }
    signOut={[MockFunction]}
    user={
      Object {
        "name": "jake",
      }
    }
    user_id={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="movies-header"
        >
          <h1
            className="movie-tracker"
          >
            movie
            <img
              alt="movie roll"
              className="main-logo"
              src="./images/movie_roll.svg"
            />
            tracker
          </h1>
          <p
            className="user-greeting"
          >
            Hello, 
            jake
          </p>
          <Link
            replace={false}
            to="/login"
          >
            <button
              className="sign-out-btn"
              onClick={[MockFunction]}
            >
              sign out
            </button>
          </Link>
          <Link
            replace={false}
            to="/favorites"
          >
            <button
              className="toggle-favorites-btn"
              onClick={[Function]}
            >
              favorites
            </button>
          </Link>
        </header>,
        <h2
          className="movie-container-label"
        >
          Trending
        </h2>,
        <div
          className="movie-container-underline"
        />,
        <div
          className="movies-container"
        >
          <Connect(MovieCard)
            movie={
              Object {
                "backdropPic": "https://image.tmdb.org/t/p/w500///",
                "genres": "drama",
                "id": 1,
                "overview": "great movie",
                "posterPic": "https://image.tmdb.org/t/p/w500///",
                "rating": 9,
                "runtime": 50,
                "title": "Titanic",
                "year": 1999,
              }
            }
            reRender={[Function]}
          />
        </div>,
      ],
      "className": "movies-section-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="movie-tracker"
            >
              movie
              <img
                alt="movie roll"
                className="main-logo"
                src="./images/movie_roll.svg"
              />
              tracker
            </h1>,
            <p
              className="user-greeting"
            >
              Hello, 
              jake
            </p>,
            <Link
              replace={false}
              to="/login"
            >
              <button
                className="sign-out-btn"
                onClick={[MockFunction]}
              >
                sign out
              </button>
            </Link>,
            <Link
              replace={false}
              to="/favorites"
            >
              <button
                className="toggle-favorites-btn"
                onClick={[Function]}
              >
                favorites
              </button>
            </Link>,
          ],
          "className": "movies-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "movie",
                <img
                  alt="movie roll"
                  className="main-logo"
                  src="./images/movie_roll.svg"
                />,
                "tracker",
              ],
              "className": "movie-tracker",
            },
            "ref": null,
            "rendered": Array [
              "movie",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "movie roll",
                  "className": "main-logo",
                  "src": "./images/movie_roll.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "tracker",
            ],
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Hello, ",
                "jake",
              ],
              "className": "user-greeting",
            },
            "ref": null,
            "rendered": Array [
              "Hello, ",
              "jake",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <button
                className="sign-out-btn"
                onClick={[MockFunction]}
              >
                sign out
              </button>,
              "replace": false,
              "to": "/login",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "sign out",
                "className": "sign-out-btn",
                "onClick": [MockFunction],
              },
              "ref": null,
              "rendered": "sign out",
              "type": "button",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <button
                className="toggle-favorites-btn"
                onClick={[Function]}
              >
                favorites
              </button>,
              "replace": false,
              "to": "/favorites",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "favorites",
                "className": "toggle-favorites-btn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "favorites",
              "type": "button",
            },
            "type": [Function],
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Trending",
          "className": "movie-container-label",
        },
        "ref": null,
        "rendered": "Trending",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "movie-container-underline",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [
              <Connect(MovieCard)
                movie={
                  Object {
                    "backdropPic": "https://image.tmdb.org/t/p/w500///",
                    "genres": "drama",
                    "id": 1,
                    "overview": "great movie",
                    "posterPic": "https://image.tmdb.org/t/p/w500///",
                    "rating": 9,
                    "runtime": 50,
                    "title": "Titanic",
                    "year": 1999,
                  }
                }
                reRender={[Function]}
              />,
            ],
            false,
          ],
          "className": "movies-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "uid1",
            "nodeType": "class",
            "props": Object {
              "movie": Object {
                "backdropPic": "https://image.tmdb.org/t/p/w500///",
                "genres": "drama",
                "id": 1,
                "overview": "great movie",
                "posterPic": "https://image.tmdb.org/t/p/w500///",
                "rating": 9,
                "runtime": 50,
                "title": "Titanic",
                "year": 1999,
              },
              "reRender": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
        ],
        "type": "div",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="movies-header"
          >
            <h1
              className="movie-tracker"
            >
              movie
              <img
                alt="movie roll"
                className="main-logo"
                src="./images/movie_roll.svg"
              />
              tracker
            </h1>
            <p
              className="user-greeting"
            >
              Hello, 
              jake
            </p>
            <Link
              replace={false}
              to="/login"
            >
              <button
                className="sign-out-btn"
                onClick={[MockFunction]}
              >
                sign out
              </button>
            </Link>
            <Link
              replace={false}
              to="/favorites"
            >
              <button
                className="toggle-favorites-btn"
                onClick={[Function]}
              >
                favorites
              </button>
            </Link>
          </header>,
          <h2
            className="movie-container-label"
          >
            Trending
          </h2>,
          <div
            className="movie-container-underline"
          />,
          <div
            className="movies-container"
          >
            <Connect(MovieCard)
              movie={
                Object {
                  "backdropPic": "https://image.tmdb.org/t/p/w500///",
                  "genres": "drama",
                  "id": 1,
                  "overview": "great movie",
                  "posterPic": "https://image.tmdb.org/t/p/w500///",
                  "rating": 9,
                  "runtime": 50,
                  "title": "Titanic",
                  "year": 1999,
                }
              }
              reRender={[Function]}
            />
          </div>,
        ],
        "className": "movies-section-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="movie-tracker"
              >
                movie
                <img
                  alt="movie roll"
                  className="main-logo"
                  src="./images/movie_roll.svg"
                />
                tracker
              </h1>,
              <p
                className="user-greeting"
              >
                Hello, 
                jake
              </p>,
              <Link
                replace={false}
                to="/login"
              >
                <button
                  className="sign-out-btn"
                  onClick={[MockFunction]}
                >
                  sign out
                </button>
              </Link>,
              <Link
                replace={false}
                to="/favorites"
              >
                <button
                  className="toggle-favorites-btn"
                  onClick={[Function]}
                >
                  favorites
                </button>
              </Link>,
            ],
            "className": "movies-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "movie",
                  <img
                    alt="movie roll"
                    className="main-logo"
                    src="./images/movie_roll.svg"
                  />,
                  "tracker",
                ],
                "className": "movie-tracker",
              },
              "ref": null,
              "rendered": Array [
                "movie",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "movie roll",
                    "className": "main-logo",
                    "src": "./images/movie_roll.svg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                "tracker",
              ],
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Hello, ",
                  "jake",
                ],
                "className": "user-greeting",
              },
              "ref": null,
              "rendered": Array [
                "Hello, ",
                "jake",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <button
                  className="sign-out-btn"
                  onClick={[MockFunction]}
                >
                  sign out
                </button>,
                "replace": false,
                "to": "/login",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "sign out",
                  "className": "sign-out-btn",
                  "onClick": [MockFunction],
                },
                "ref": null,
                "rendered": "sign out",
                "type": "button",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <button
                  className="toggle-favorites-btn"
                  onClick={[Function]}
                >
                  favorites
                </button>,
                "replace": false,
                "to": "/favorites",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "favorites",
                  "className": "toggle-favorites-btn",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "favorites",
                "type": "button",
              },
              "type": [Function],
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Trending",
            "className": "movie-container-label",
          },
          "ref": null,
          "rendered": "Trending",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "movie-container-underline",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [
                <Connect(MovieCard)
                  movie={
                    Object {
                      "backdropPic": "https://image.tmdb.org/t/p/w500///",
                      "genres": "drama",
                      "id": 1,
                      "overview": "great movie",
                      "posterPic": "https://image.tmdb.org/t/p/w500///",
                      "rating": 9,
                      "runtime": 50,
                      "title": "Titanic",
                      "year": 1999,
                    }
                  }
                  reRender={[Function]}
                />,
              ],
              false,
            ],
            "className": "movies-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "uid1",
              "nodeType": "class",
              "props": Object {
                "movie": Object {
                  "backdropPic": "https://image.tmdb.org/t/p/w500///",
                  "genres": "drama",
                  "id": 1,
                  "overview": "great movie",
                  "posterPic": "https://image.tmdb.org/t/p/w500///",
                  "rating": 9,
                  "runtime": 50,
                  "title": "Titanic",
                  "year": 1999,
                },
                "reRender": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
          ],
          "type": "div",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`MovieDisplay Should match the snapshot if isLoading is true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieDisplay
    isLoading={true}
    loadMovies={
      [MockFunction] {
        "calls": Array [
          Array [
            "https://api.themoviedb.org/3/movie/popular?api_key=da90047b6c1d3526d4b04666a1b64a0d&language=en-US&page=1&region=US",
          ],
          Array [
            "https://api.themoviedb.org/3/movie/popular?api_key=da90047b6c1d3526d4b04666a1b64a0d&language=en-US&page=1&region=US",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    movies={
      Array [
        Object {
          "backdropPic": "https://image.tmdb.org/t/p/w500///",
          "genres": "drama",
          "id": 1,
          "overview": "great movie",
          "posterPic": "https://image.tmdb.org/t/p/w500///",
          "rating": 9,
          "runtime": 50,
          "title": "Titanic",
          "year": 1999,
        },
      ]
    }
    signOut={[MockFunction]}
    user={
      Object {
        "name": "jake",
      }
    }
    user_id={2}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
